### ISSUE TITLE
CompilerServer: server failed due to Xunit Analyzers 1.14

### ISSUE THREAD

**Issue by Tornhoof** [xunit/xunit#2943]
After update  to XUnit Analyzers 1.14 Azure Devops Build outputs the following during the build (for every test project):

```ini
  CompilerServer: server failed - server rejected the request due to analyzer / generator issues 'analyzer assembly 'C:\Users\BuildUser\.nuget\packages\xunit.analyzers\1.14.0\analyzers\dotnet\roslyn4.8\cs\xunit.analyzers.dll' has MVID '3e2596a5-a0c9-41a0-b5cf-9fe156e7ea5f' 
  but loaded assembly 'C:\Users\BuildUser\AppData\Local\Temp\VBCSCompiler\AnalyzerAssemblyLoader\c7412f87926d47a2a5eb7eda0608e533\10\xunit.analyzers.dll' has MVID 'b1a42a47-0126-4f64-b679-249d385e60a7', 
  analyzer assembly 'C:\Users\BuildUser\.nuget\packages\xunit.analyzers\1.14.0\analyzers\dotnet\roslyn4.8\cs\xunit.analyzers.fixes.dll' has MVID '2dec3629-615b-446a-b634-e65e8cb5dffb' 
  but loaded assembly 'C:\Users\BuildUser\AppData\Local\Temp\VBCSCompiler\AnalyzerAssemblyLoader\c7412f87926d47a2a5eb7eda0608e533\11\xunit.analyzers.fixes.dll' has MVID 'bfdb98a4-35e9-4e80-97d9-33b9157e3ce3'' - ec95a20c-48f8-44d0-a766-e8629738d172
```
I've noticed 1.14 now contains a folder 'cs' in analyzers/dotnet, you've added the cs folder here: https://github.com/xunit/xunit.analyzers/commit/9cbb182d2e95088f6015695e80e3d90fa02265e2

I found a related issue here: https://github.com/microsoft/vs-mef/issues/204 where they point to version issues

So what happens according to the message above is, that roslyn wants to load the roslyn4.8 dll (with module id 3e2596a5-a0c9-41a0-b5cf-9fe156e7ea5f) but the one from 'cs' is already loaded  with module id (b1a42a47-0126-4f64-b679-249d385e60a7).

I'm not sure about the impact of the message, the build is slower, prior to updates the build took ~50s, now it's ~150s, probably because it needs to restart the compiler server to get rid of the other dll?

As for repro, a pretty much empty Test Project in VS does the trick
```csharp
using Xunit;

namespace TestProject1
{
    public class UnitTest1
    {
        [Fact]
        public void Test1()
        {

        }
    }
}
```

```ini
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>net8.0</TargetFramework>
    <Nullable>enable</Nullable>
  </PropertyGroup>
  
  <ItemGroup>
    <PackageReference Include="Microsoft.CodeCoverage" Version="17.10.0" />
    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.10.0" />
    <PackageReference Include="xunit" Version="2.8.1" />
    <PackageReference Include="xunit.assert" Version="2.8.1" />
    <PackageReference Include="xunit.runner.visualstudio" Version="2.8.1">
      <PrivateAssets>all</PrivateAssets>
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
    </PackageReference>
  </ItemGroup>
</Project>

```

Then run from command line: 
- dotnet restore
-  msbuild TestProject1.sln /p:configuration="release"

The output will now contain the compiler server error message, at the end of step 'CoreCompile'
You also see a noticeable pause after the error message. If you downgrade to 2.8.0, both the error message and pause are gone.

**Comment by bradwilson** [comment]
The reason there is a version in the `cs` folder is that the analyzers were no longer working in "old" style CSPROJ files (meaning, not the new SDK-based CSPROJ files) because they don't understand the Roslyn folders.

The only way I know of to "fix" this would be to stop shipping Roslyn-specific versions, and thus everybody will get whatever is available via Roslyn 3.11.

**Comment by Tornhoof** [comment]
Or splitting the legacy 'cs' one into a different nuget package, 'xunit.analyzers.legacy' or something, which would be a breaking change.  This kinda sounds like something the msbuild or roslyn team needs to fix, i find it surprising that the legacy folder is apparently evaluated first.

**Comment by bradwilson** [comment]
Agreed that this does not appear to be our bug.

That said, a separate NuGet package is a non-starter. That would put the onus on the developer to (a) know which package is correct, and (b) add it; compare that to today where 99%+ of people just import the `xunit` package and be done with it.

**Comment by sharwell** [comment]
> The only way I know of to "fix" this would be to stop shipping Roslyn-specific versions, and thus everybody will get whatever is available via Roslyn 3.11.

For analyzers, this is generally what I would recommend. A lightup layer that uses reflection can fill in the functionality gaps where needed. The complexity of this approach depends on how frequently the library depends on newer APIs than are available in the version referenced at build time.

**Comment by bradwilson** [comment]
Available in `1.15.0-pre.6` https://xunit.net/docs/using-ci-builds
